Spark

taragu opened the pull request:
Added Spark standalone, streaming examples, and integration with Cloud Pubsub

taragu commented:
@lesv have you found any problems with my code? Thank you!

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+
+Before you run this code sample, you'll need to complete the following tasks:
+
+1. [Create a Cloud Bigtable cluster][create-cluster]. Be sure to note the
```
Links are done by [text]\(link) There should be links to documentation or other explanation here.

taragu replied:
I was able to open the links on my computer by clicking "Create a Cloud Bigtable cluster" from here https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-pubsub/README.md. Do I still need to change it to [text](link)?

lesv replied:
No -- it's a part of Markdown I guess I missed. :(

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+
+1. [Create a Cloud Bigtable cluster][create-cluster]. Be sure to note the
+project ID.
+2. [Create a service account and a JSON key file][json-key].
```
Also here.

lesv commented:
@maxluebbe PTAL
@hegemonic PTAL

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+
+You can use the Spark-Cloud Pubsub connector in two ways. If you would like to use it without any modification, you can download a pre-built jar from a GCS storage bucket with the following command:
+
+    $ gsutil cp gs://cloud-bigtable-examples/spark-cloud-pubsub-connector_2.10-0.0.jar PATH/TO/SAVE/THE/FILE
```
@sduskis Is this really where / how we wish to distribute this?  Shouldn't we add it to our mvn repo?

lesv replied:
@sduskis @taragu @hegemonic After having a conversation in DevRel - we are recommending that we do not put jars in buckets, but in the GitHub Releases.

taragu replied:
@lesv How do I put the jar in the releases?

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+# Example: Integrating Spark Streaming with Cloud Pubsub
+
```
Please put a Table of contents

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+into a GCS bucket, log in to the master VM, then download the it from the GCS
+bucket. You can also download the pre-built connector instead of building it on your machine with the following command:
+
+    $ gsutil cp gs://cloud-bigtable-examples/spark-cloud-pubsub-connector_2.10-0.0.jar PATH/TO/SAVE/THE/FILE
```
From the releases folder.

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+
+First, create a new topic in the Cloud Pubsub web UI. Please note the topic name.
+
+Next, open two terminals (A and B). In terminal A, we run the message producer; in terminal B, we run the message processor. 
```
shell / terminal windows / shell tabs

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+“Building the code sample”, or download a pre-compiled jar with the following
+command:
+
+	$ gsutil cp gs://cloud-bigtable-examples/cloud-pubsub-receiver_2.10-0.0.jar .
```
github/gcp/cloud-bigtable-examples/releases (figure out real URL)

lesv commented on scala/spark-pubsub/README.md:
```diff
@@ -0,0 +1,161 @@
+(Still in terminal B) We also need to Cloud Pubsub API jar as well as the Spark-Cloud Pubsub connector on Spark's classpath in order to call their API in runtime. Download the Cloud Pubsub API jar and the connector jar with the following commands:
+
+    $ wget http://central.maven.org/maven2/com/google/apis/google-api-services-pubsub/v1-rev2-1.20.0/google-api-services-pubsub-v1-rev2-1.20.0.jar
+    $ gsutil cp gs://cloud-bigtable-examples/spark-cloud-pubsub-connector_2.10-0.0.jar .
```
ditto

lesv commented:
The JAR shouldn't be submitted here.

lesv commented on scala/spark-pubsub/cloud-pubsub-producer/cloudpubsubproducer.sbt:
```diff
@@ -0,0 +1,17 @@
+      "org.apache.hbase" % "hbase-server" % "1.0.1", 
+      "org.apache.hbase" % "hbase-common" % "1.0.1",
+      "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+      "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317",
```
Highly unlikely this is correct.

taragu replied:
What do you mean by submitting the jar here?

lesv replied:
7.1.3 is likely the correct version -- bdutil is incorrect if it's using this.

lesv commented on scala/spark-pubsub/cloud-pubsub-producer/src/main/scala/CloudPubsubProducer.scala:
```diff
@@ -0,0 +1,49 @@
+      throw new Exception("Please enter project name, topic name,"
+        + " and input file name as arguments. The project and "
+        + "topic must exist. Example: sbt \"project cloud-pubsub-"
+        + "producer\" \"run sduskis-hello-shakespear cmd_line_1 "
```
No way this is right.  Never have our names on stuff.

lesv commented on scala/spark-pubsub/cloud-pubsub-receiver/cloudpubsubreceiver.sbt:
```diff
@@ -0,0 +1,16 @@
+      "org.apache.hbase" % "hbase-server" % "1.0.1", 
+      "org.apache.hbase" % "hbase-common" % "1.0.1",
+      "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+      "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317",
```
7.1.3 alpn is more likely.

taragu replied:
Bdutil uses 7.0.0: https://github.com/GoogleCloudPlatform/bdutil/blob/f1b5b36bae4bd8b5abb09d94d3575980b307755d/extensions/bigtable/bigtable_env.sh#L40
Should I use 7.1.3 instead even though it won't be in the VMs?

lesv replied:
@sduskis @taragu bdutil is wrong then.  alpn-boot needs to match the version of java you are using -- http://www.eclipse.org/jetty/documentation/current/alpn-chapter.html#alpn-versions

lesv commented on scala/spark-pubsub/spark-cloud-pubsub-connector/sparkcloudpubsubconnector.sbt:
```diff
@@ -0,0 +1,16 @@
+		    "org.apache.hbase" % "hbase-server" % "1.0.1", 
+		    "org.apache.hbase" % "hbase-common" % "1.0.1",
+		    "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+		    "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317",
```
7.1.3 is more likely

lesv commented on scala/spark-standalone-wordcount/README.md:
```diff
@@ -0,0 +1,20 @@
+
+Spark: https://spark.apache.org/
+
+Please note that we encourage users to develop programs on their local machine, not on the VMs. User 
```
We probably don't need this sentence.  The rest of the paragraph is fine.

lesv commented on scala/spark-standalone-wordcount/README.md:
```diff
@@ -0,0 +1,20 @@
+	 $ ./bdutil -e hadoop2_env.sh -e extensions/spark/spark_env.sh -e extensions/bigtable/bigtable_env.sh -e [path/to/config/file.sh] -f deploy
+
+Please go to the official documentation for specific instructions:
+TODO include link
```
We should include more.

taragu replied:
What should we include other than the link to the official documentation?

lesv commented on scala/spark-standalone-wordcount/src/main/scala/WordCount.scala:
```diff
@@ -0,0 +1,134 @@
+    val conn = ConnectionFactory.createConnection(
+      confBroadcast.value.value)
+
+    // create new table if it's not already existed
```
Create a new table if it doesn't already exist.

lesv commented on scala/spark-standalone-wordcount/src/main/scala/WordCount.scala:
```diff
@@ -0,0 +1,134 @@
+      partition => {
+        val config = confBroadcast.value.value
+	val conn1 = ConnectionFactory.createConnection(config)
+        val tableName1 = TableName.valueOf(name)
```
indentation

taragu replied:
Hmm I don't know why it shows up like this in the diff. It looks fine from here: https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/src/main/scala/WordCount.scala.

lesv replied:
Look for special characters in the file.  If you are happy w/ it, then I'll be.

lesv commented on scala/spark-standalone-wordcount/src/main/scala/WordCount.scala:
```diff
@@ -0,0 +1,134 @@
+	val conn1 = ConnectionFactory.createConnection(config)
+        val tableName1 = TableName.valueOf(name)
+        val mutator = conn1.getBufferedMutator(tableName1)
+	try {
```
indentation ???

lesv commented on scala/spark-standalone-wordcount/wordcount.sbt:
```diff
@@ -0,0 +1,13 @@
+		    "org.apache.hbase" % "hbase-server" % "1.0.1.1", 
+		    "org.apache.hbase" % "hbase-common" % "1.0.1.1",
+		    "org.apache.hbase" % "hbase-annotations" % "1.0.1.1",
+		    "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317")
```
7.1.3 ???

lesv commented on scala/spark-streaming-wordcount/README.md:
```diff
@@ -0,0 +1,21 @@
+
+Spark Streaming: http://spark.apache.org/docs/latest/streaming-programming-guide.html
+
+Please note that we encourage users to develop programs on their local machine, not on the VMs. User 
```
Remove the "please note" sentence.  I think I gave that to you as advice on how to write, not to put it in the readme -- please remove from all readme's.

lesv commented on scala/spark-streaming-wordcount/streamingexample.sbt:
```diff
@@ -0,0 +1,15 @@
+		    "org.apache.hbase" % "hbase-server" % "1.0.1", 
+		    "org.apache.hbase" % "hbase-common" % "1.0.1",
+		    "org.apache.hbase" % "hbase-annotations" % "1.0.1",
+		    "org.mortbay.jetty.alpn" % "alpn-boot" % "7.0.0.v20140317")
```
7.1.3 ??

lesv commented:
Let me know when all of these are done.

taragu commented:
@lesv bdutil uses ALPN 7.0.0: https://github.com/GoogleCloudPlatform/bdutil/blob/f1b5b36bae4bd8b5abb09d94d3575980b307755d/extensions/bigtable/bigtable_env.sh#L40
Should we use 7.1.3 instead even though it won't be in the VMs?

taragu commented:
@lesv I've addressed your comments. The indentation looks messed up in the diff but looks fine here: https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-streaming-wordcount/src/main/scala/StreamingExample.scala
https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/src/main/scala/WordCount.scala

Also, you mentioned I should include more in the word count example README (https://github.com/taragu/cloud-bigtable-examples/blob/spark/scala/spark-standalone-wordcount/README.md) besides the link to the official doc. What else should I include?

sduskis commented:
I find that indentation issues happen when there are tabs in the document.

lesv commented on scala/spark-standalone-wordcount/src/main/scala/WordCount.scala:
```diff
@@ -0,0 +1,134 @@
+    // broadcast a serialized config object allows us to use the
+    // same conf object among the driver and executors
+    val confBroadcast = sc.broadcast(
+      new SerializableWritable(hbaseConfig))
```
Should this be indent by 4?  I don't recall seeing anything in the guide, but most of our guides double the spacing for a carry on line.

lesv commented:
Tara -- Look at Solomon's advice and remove any tabs or hidden characters.  Consider using 4 space indents for carry on lines.

I need to check w/ @hegemonic as we may just do a simple doc for this and have the bulk of the text in the various readme's.

taragu commented:
@lesv I've removed the tabs. As we discussed earlier, I didn't use 4 spaces because the style guide used 2 spaces: https://sites.google.com/a/google.com/scala/coding/scala-style-for-new-developers?pli=1. Is there anything else?

lesv commented:
Did you just strike the alpn-boot?  If so, does it work?

taragu commented:
Yes---I deleted alpn in the build files because it was only used in runtime. We don't need it to compile the code.

lesv commented:
LGTM -- please make sure to add it to the README about running, but lets start a new PR for that.
